packages/asset-swapper/src/utils/market_operation_utils/router_test_data_sampler.ts

```diff
@@ -0,0 +1,57 @@
+import { ChainId } from '@0x/contract-addresses';
+import { OptimizerCapture } from '@0x/neon-router';
+import { CreateBucketCommand, PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
+
+const SAMPLE_THRESHOLD = 0.05;
+const UPLOAD_SIZE = 1000;
+const REGION = process.env.AWS_S3_REGION ? process.env.AWS_S3_REGION : 'us-west-1';
+
+export class TestDataSampler {
+    private static _instance: TestDataSampler;
+    private _routes: OptimizerCapture[] = [];
+    private readonly _chainId;
+    public static getInstance(chainId: ChainId): TestDataSampler {
+        // singleton implementation
+        if (!TestDataSampler._instance) {
+            TestDataSampler._instance = new TestDataSampler(chainId);
+        }
+        return TestDataSampler._instance;
+    }
+
+    public sampleRoute(route: OptimizerCapture): void {
+        if (Math.random() < SAMPLE_THRESHOLD) {
+            this._routes.push(route);
+
+            if (this._routes.length > UPLOAD_SIZE) {
+                const toUpload = this._routes;
+                this._routes = [];
+                const file = `${this._chainId}-${Date.now()}`;
```

Comment:
This would be the folder path, not the file right? If so the timestamp should go into the key otherwise we'll end up creating thousands of buckets if I'm not mistaken 🤔 

I would create a bucket like `0x-router-test-data` and this would be the bucket.

Then the key would be something like`${this._chainId}/sample-${Date.now()}.json`, that should put them into "subfolder" in s3 IIRC

Reply:
Yeah the previous approach generates a lot of buckets. I kinda forgot the number of buckets allowed is limited. Good pick 👍